polygon/heimdall/entity_fetcher.go

```diff
@@ -0,0 +1,133 @@
+package heimdall
+
+import (
+	"cmp"
+	"context"
+	"fmt"
+	"slices"
+	"time"
+
+	"github.com/ledgerwatch/log/v3"
+)
+
+type entityFetcher interface {
+	FetchLastEntityId(ctx context.Context) (uint64, error)
+	FetchEntitiesRange(ctx context.Context, idRange EntityIdRange) ([]any, error)
+}
+
+type genericEntityFetcher[TEntity entity] struct {
+	name string
+
+	fetchLastEntityId func(ctx context.Context) (int64, error)
+	fetchEntity       func(ctx context.Context, id int64) (*TEntity, error)
+	fetchEntitiesPage func(ctx context.Context, page uint64, limit uint64) ([]*TEntity, error)
+	getStartBlockNum  func(entity *TEntity) uint64
+
+	logger log.Logger
+}
+
+func newGenericEntityFetcher[TEntity entity](
```

Comment:
unit tests?

Reply:
The logic is copy-pasted from heimdall.go/store.go. As I refactor heimdall.go/store.go I was planning its tests onto the new stuff. Currently we have 2 things in parallel.

When it comes to new unit tests, please be more specific. I prefer to not add tests for trivial code only to please the coverage report. Please tell me which use cases you want covered, and I'll work on it.

Reply:
IMO we should have tests for `Scraper.syncEntity` and `genericEntityFetcher.FetchAllEntities`. The rest is trivial, and tests would not be beneficial.

Scraper.syncEntity is doing what previously was a responsibility of heimdall.OnXXXEvent methods. We have a test for TestOnMilestoneEvent. I will look into it and try to port to the new code.

genericEntityFetcher.FetchAllEntities is doing what previously was a responsibility of heimdall.batchFetchCheckpoints. We don't have a test for it specifically, but we have some tests for TestFetchCheckpointsXXX. I'll look if it can be reused, or if we need a new tests.

We should establish some guidelines about when to NOT add unit tests.

Reply:
FetchEntitiesRange - needs testing for 2 cases: happy path & conditional branch for when it should FetchAllEntities
FetchEntitiesRange - needs test for happy path
FetchLastEntityId - needs test for happy path

you say this "trivial code" but it is not - if you keep using this as an excuse you will end up with 20 "trivial" components which have 0 test coverage

Reply:
my guidelines are:
- always have unit tests for exported (functions)
- cover all branching of a func
- private funcs dont need to be unit tested - they can be tested via public API flow